ml-fundamentals/ml-testing.md

```diff
@@ -0,0 +1,213 @@
+```
+
+To test different inputs and expected outputs automatically, use parametrize
+
+```python
+@pytest.mark.parametrize('orig_height, orig_width, expected_height, expected_width',
+                         [
+                             # smaller than target
+                             (10, 10, 20, 20),
+                             # larger than target
+                             (20, 20, 10, 10),
+                             # wider than target
+                             (10, 20, 10, 10)
+                         ])
+def test_resize_image_generates_the_correct_size(orig_height, orig_width, expected_height, expected_width):
+  # Arrange
+  original_image = np.ones((orig_height, orig_width, 2, 3))
+
+  # act
+  resized_image = utils.resize_image(original_image, expected_height, expected_width)
+
+  # assert
+  resized_image.shape[:2] = (expected_height, expected_width)
+```
+
+## Model load or predict
+
+When **unit** testing we should mock model load and model predictions similarly to mocking file access.
+
+There may be cases when you want to load your model to do smoke tests, or integration tests.
+
+Since these will often take a bit longer to run it's important to be able to separate them from unit tests so that the developers on the team can still run unit tests as part of their test driven development.
+
+One way to do this is using marks
+
+```python
+@pytest.mark.longrunning
+def test_integration_between_two_systems():
+  # this might take a while
+```
+
+Run all tests that are not marked longrunning
+
+```bash
+pytest -v -m "not longrunning"
+```
+
+### Basic Unit Tests for ML Models
+
+ML unit tests are not intended to check the accuracy or performance of a model. Unit tests for an ML model is for code quality checks - for example:
+
+* Does the model accept the correct inputs and produce the correctly shaped outputs?
+* Do the weights of the model update when running `fit`?
+
+To do this, the ML model tests do not strictly follow all of the best practices of standard Unit tests - not all outside calls are mocked. These tests are much closer to a [narrow integration test](https://martinfowler.com/bliki/IntegrationTest.html).
+However, the benefits of having simple tests for the ML model help to stop a poorly configured model from spending hours in training, while still producing poor results.
+
+Examples of how to implement these tests (for Deep Learning models) include:
+
+* Build a model and compare the shape of input layers to that of an example source of data. Then, compare the output layer shape to the expected output.
+* Initialize the model and record the weights of each layer. Then, run a single epoch of training on a dummy data set, and compare the weights of the "trained model" - only check if the values have changed.
+* Train the model on a dummy dataset for a single epoch, and then validate with dummy data - only validate that the prediction is formatted correctly, this model will not be accurate.
```

Comment:
What do you think about having a test to verify the correct results on the simplest possible case? For example, in NER, whether the model detects a person in the sentence "My name is David".